logger/src/main/kotlin/org/jetbrains/bsp/bazel/logger/BspClientTestNotifier.kt

```diff
@@ -52,6 +54,20 @@ class BspClientTestNotifier(private val bspClient: BuildClient, private val orig
     bspClient.onBuildTaskFinish(taskFinishParams)
   }
 
+  /**
+   * Notifies the client of an available coverage report file.
+   *
+   * @param taskId         TaskId of the task that produced the coverage report
+   * @param lcovReportPath full path path to the coverage report file
+   */
+  fun sendCoverageReport(taskId: TaskId, lcovReportPath: String) {
+    val taskProgressParams = TaskProgressParams(taskId)
```

Comment:
hmm, I think it's a very good question whether this should be a task progress notification, a separate task or maybe a completely separate notification (like sending diagnostics). WDYT @jastice?

Reply:
Yeah good point.  My thought process in putting this under progress and not under its own task is that this this is one output file among potentially several others within the overall test run task, and not inherently a task on its own.

However that's a good point that we could potentially leverage diagnostics here:
- bazel-bsp could handle parsing of the coverage file
- conversion to document ranges (basically 1 diagnostic for every line of each file in the run)
- sending those to the client as diagnostics which include execution counts per line
- the client would then need to have some handling that treats these diagnostics as coverage highlighting instead of regular diagnostics

I think it might get a bit more involved, maybe we could start off with the initial step of exposing a path to the coverage file as it is now, and looking at diagnostics as a next step?

Reply:
Agree, ideally we'd use diagnostics here, maybe with additional `CoverageDiagnosticData`. This could include the file as first step

It seems a bit off to use progress, unless there is actual progress to report?

Reply:
Ok, I can try switching it over to a diagnostic.  Only thing is that publishDiagnostics doesn't seem great for sending a single file path either, as it requires a TextDocumentIdentifier in the notification, but the coverage report is a target level output that doesn't belong to a particular location in any one document.  Maybe an optional field on the test report could be another option.

Reply:
For the first step of just returning the file, I tried to switch this over to diagnostics, but the part about assigning it to a TextDocumentIdentifier feels a bit forced, since it's more of a top-level output of the test target and not associated with any one text document.  Also, even if we do pick a document to add there, the user would see a uri to a file that is not really human readable and is meant for processing by the client.

The _contents_ could potentially be parsed and converted to something diagnostic-like, but earlier in the thread we had discussed that at least for now, the client can parse it.

It feels like data that should be more associated with a task than with a document, either as its own task, progress on the target task (even though it's not numerical progress, "this new output is available" is still a form of progress), or another field in the test finish data.  Thoughts on any of those?

Reply:
What I meant with diagnostics was not necessarily reusing the existing diagnostics notification, but rather making a new one, completely separate from the task notification system ;)

Reply:
Oh ok, that makes more sense :)  So maybe a new notification called something like `build/publishOutput` and params like this?  It can be used for similar cases where there is a need to send an output from the build to the client.

```
export interface PublishOutputParams {
  /** The build target where the output originates.
  buildTarget: BuildTargetIdentifier;

  /** The request id that originated this notification. */
  originId?: OriginId;

  /** Kind of data to be included here. */
  dataKind: string,

  /** Data matching dataKind above. */
  data: Any;
}
```

Reply:
That sounds reasonable, if we have other good use cases for such a notification (test report files? build artifacts?). Otherwise I might just go with a dedicated one for this use-case.

Reply:
>if we have other good use cases for such a notification (test report files? build artifacts?)

I don't have any immediately off the top of my head, but yeah I was thinking other cases like those might come up in the future and it would be easier to support them if the notification is more general.

Reply:
I made the update above.  For the Client, it looks like this is the first notification type outside the existing protocol, so I had to create a JoinedBuildClient, similar in approach to the JoinedBuildServer - as I had to update this in a few other places, let me know if I did anything incorrect there.

I made the params essentially the same as above, and notification `build/publishOutput`.  I was able to test this with our client and received the notification.

Let me know if you all have any other feedback on this.